pkg/resource/autonaming/pattern.go

```diff
@@ -0,0 +1,183 @@
+// These are all expressions that can be resolved at the startup time because they don't depend
+// on the resource URN.
+type stackPatternEval struct {
+	org            string
+	proj           string
+	stack          string
+	getConfigValue func(key string) (string, error)
+}
+
+// newStackPatternEval creates a new stack pattern evaluator based on the given stack and configuration.
+func newStackPatternEval(s backend.Stack, cfg *backend.StackConfiguration, decrypter config.Decrypter,
+) *stackPatternEval {
+	org := "default"
+	if cs, ok := s.(httpstate.Stack); ok {
+		org = cs.OrgName()
+	}
+	projName, _ := s.Ref().Project()
+	projStr := projName.String()
+	stackStr := s.Ref().Name().String()
+	getConfigValue := func(key string) (string, error) {
+		c, ok, err := cfg.Config.Get(config.MustMakeKey(projStr, key), true)
+		if err != nil {
+			return "", fmt.Errorf("failed to get config value for key %q: %w", key, err)
+		}
+		if !ok {
+			return "", fmt.Errorf("no value found for key %q", key)
+		}
+		v, err := c.Value(decrypter)
+		if err != nil {
+			return "", fmt.Errorf("failed to decrypt value for key %q: %w", key, err)
+		}
+		return v, nil
+	}
+	return &stackPatternEval{
+		org:            org,
+		proj:           projStr,
+		stack:          stackStr,
+		getConfigValue: getConfigValue,
+	}
+}
+
+// resolveStackExpressions resolves the org, project, stack, and config expressions in the given pattern.
+func (e *stackPatternEval) resolveStackExpressions(pattern string) (string, error) {
+	// Replace ${org}, ${project}, ${stack} with values from context
+	pattern = strings.ReplaceAll(pattern, "${org}", e.org)
+	pattern = strings.ReplaceAll(pattern, "${project}", e.proj)
+	pattern = strings.ReplaceAll(pattern, "${stack}", e.stack)
+
+	// Replace ${config.key} with config values
+	configRegex := regexp.MustCompile(`\${config\.([^}]+)}`)
+	var configErr error
+	pattern = configRegex.ReplaceAllStringFunc(pattern, func(match string) string {
+		key := configRegex.FindStringSubmatch(match)[1]
+		v, err := e.getConfigValue(key)
+		if err != nil {
+			configErr = err
+			return ""
+		}
+		return v
+	})
+	if configErr != nil {
+		return "", configErr
+	}
+	return pattern, nil
+}
+
+// generateName generates a final proposed name based on the configured pattern, the resource URN, and random seed.
+// Note that the pattern is expected to have already had stack-level expressions like ${org}, ${project}, ${stack},
+// and ${config.key} resolved before passing to this function.
+func generateName(pattern string, urn urn.URN, randomSeed []byte) (string, bool) {
+	// Replace ${name} with the logical name
+	result := strings.ReplaceAll(pattern, "${name}", urn.Name())
+	hasRandom := false
+
+	// Create a random number generator with the given seed. If no seed is provided, use a default
+	// random number generator.
+	var random *frand.RNG
+	if len(randomSeed) == 0 {
+		random = frand.New()
+	} else {
```

Comment:
Feel like rather than this `if/else` you could just generate a random seed if none is supplied then unconditionally source a generator with that.

Reply:
What would be the benefit of doing so?

Again, I took it from `resource.NewUniqueName()`, maybe I should refactor that piece of code into a reusable function instead of copy-pasting.